### PR TITLE
MGMT-18561: MGMT-18562: Add kube api support for adding per cluster mirror registry in AgentClusterInstall

### PR DESCRIPTION
**Pull request by eliorerz**
## List all the issues related to this PR

- [x] New Feature 
- [ ] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [x] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [x] Waiting for CI to do a full test run
- [x] Manual (Elaborate on how it was tested)
- [ ] No tests needed

/cc @eliorerz

### CODE REVIEW

**Comment by CrystalChun** [comment]
Retested and verified it's pulling from the mirror registry 👍 
```
  Warning  Failed          1s (x2 over 15s)   kubelet            Failed to pull image "quay.io/cchun/testing@sha256:e1f8ac471c4cd63333590309829aa8f1210a62571179d39fc6989ab6a9c09e2e": (Mirrors also failed: [10.1.178.25:5000/cchun/testing@sha256:e1f8ac471c4cd63333590309829aa8f1210a62571179d39fc6989ab6a9c09e2e: pinging container registry 10.1.178.25:5000: Get "https://10.1.178.25:5000/v2/": http: server gave HTTP response to HTTPS client]): quay.io/cchun/testing@sha256:e1f8ac471c4cd63333590309829aa8f1210a62571179d39fc6989ab6a9c09e2e: reading manifest sha256:e1f8ac471c4cd63333590309829aa8f1210a62571179d39fc6989ab6a9c09e2e in quay.io/cchun/testing: unauthorized: access to the requested resource is not authorized
```

File: internal/bminventory/inventory.go [hunk 1]
```diff
@@ -1279,7 +1281,7 @@ func (b *bareMetalInventory) integrateWithAMSClusterPreInstallation(ctx context.
 	return nil
 }
 
-func (b *bareMetalInventory) InstallClusterInternal(ctx context.Context, params installer.V2InstallClusterParams) (*common.Cluster, error) {
+func (b *bareMetalInventory) InstallClusterInternal(ctx context.Context, params installer.V2InstallClusterParams, mirrorRegistryConfiguration *v1beta1.MirrorRegistryConfiguration) (*common.Cluster, error) {
```
Comment (carbonin) [inline internal/bminventory/inventory.go:1284]: Looks like this was added and isn't used, can we remove it?

File: internal/bminventory/inventory.go [hunk 2]
```diff
@@ -1670,6 +1672,13 @@ func (b *bareMetalInventory) UpdateClusterInstallConfigInternal(ctx context.Cont
 			return common.NewApiError(http.StatusBadRequest, err)
 		}
 
+		if mirrorregistries.IsMirrorConfigurationSet(mirrorRegistryConfiguration) {
```
Comment (carbonin) [inline internal/bminventory/inventory.go:1675]: Why do this check?

File: internal/bminventory/inventory_test.go [hunk 3]
```diff
@@ -13188,6 +13178,65 @@ var _ = Describe("TestRegisterCluster", func() {
...
+			mockClusterRegisterSuccess(true)
+			mockAMSSubscription(ctx)
+			conf, _ := getMirrorRegistryConfigurations(getSecureRegistryToml(), mirrorRegistryCertificate)
+			params := getClusterCreateParams()
+			c, err := bm.RegisterClusterInternal(ctx, nil, conf, installer.V2RegisterClusterParams{NewClusterParams: params})
+			Expect(err).ShouldNot(HaveOccurred())
+
+			var clusterObj common.Cluster
+			Expect(db.First(&clusterObj, "id = ?", c.ID).Error).ShouldNot(HaveOccurred())
+			Expect(clusterObj.MirrorRegistryConfiguration).ShouldNot(BeNil())
```
Comment (carbonin) [inline internal/bminventory/inventory_test.go:13236]: Is there anything about this content you can verify?

File: internal/bminventory/inventory_test.go [hunk 4]
```diff
@@ -13188,6 +13178,65 @@ var _ = Describe("TestRegisterCluster", func() {
 		}
 	})
 
+	Context("Cluster Mirror Registry", func() {
```
Comment (carbonin) [inline internal/bminventory/inventory_test.go:13181]: Can you add a test for the infraenv functions as well as the cluster update one?

File: internal/bminventory/inventory.go [hunk 5]
```diff
@@ -1647,7 +1649,7 @@ func (b *bareMetalInventory) GetArchitecturesSupportLevelListInternal(_ context.
 	return featuresupport.GetCpuArchitectureSupportList(params.OpenshiftVersion), nil
 }
 
-func (b *bareMetalInventory) UpdateClusterInstallConfigInternal(ctx context.Context, params installer.V2UpdateClusterInstallConfigParams) (*common.Cluster, error) {
+func (b *bareMetalInventory) UpdateClusterInstallConfigInternal(ctx context.Context, params installer.V2UpdateClusterInstallConfigParams, mirrorRegistryConfiguration *v1beta1.MirrorRegistryConfiguration) (*common.Cluster, error) {
```
Comment (carbonin) [inline internal/bminventory/inventory.go:1652]: Why have this as a parameter to install config and also to the cluster itself. I would think it would be better to keep it in one place (just in the cluster).

File: internal/common/db.go [hunk 6]
```diff
@@ -537,3 +589,27 @@ func CloseDB(db *gorm.DB) {
...
+	err := json.Unmarshal(data, &config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal MirrorRegistryConfiguration: %w", err)
+	}
+
+	return &config, nil
+}
+
+// ConvertMirrorRegistryConfigToBytes converts a MirrorRegistryConfiguration struct to a byte slice
+func ConvertMirrorRegistryConfigToBytes(config *v1beta1.MirrorRegistryConfiguration) ([]uint8, error) {
```
Comment (carbonin) [inline internal/common/db.go:607]: Does this need to be exported?

File: internal/common/db.go [hunk 7]
```diff
@@ -191,6 +218,9 @@ type InfraEnv struct {
 	// Json formatted string containing internal overrides for the default ignition config.
 	// This is used for adding ironic ignition config to the assisted ignition config
 	InternalIgnitionConfigOverride string `json:"internal_ignition_config_override,omitempty"`
+
+	// A JSON blob in which holds a mirror registry configurations if set
+	MirrorRegistryConfiguration []uint8 `gorm:"type:BYTEA"`
```
Comment (carbonin) [inline internal/common/db.go:223]: Why byte and not just serialized json?

File: api/hiveextension/v1beta1/agentclusterinstall_types.go [hunk 8]
```diff
@@ -505,3 +512,20 @@ type ManifestsConfigMapReference struct {
...
+// MirrorRegistryConfigMapReference contains reference to a ConfigMap for mirror registry
+type MirrorRegistryConfigMapReference struct {
+	// Name is the name of the ConfigMap that this refers to
+	Name string `json:"name"`
+	// Namespace of the ConfigMap
+	Namespace string `json:"namespace"`
+}
+
+// MirrorRegistryConfiguration holds the given mirror registry configuration
+type MirrorRegistryConfiguration struct {
```
Comment (carbonin) [inline api/hiveextension/v1beta1/agentclusterinstall_types.go:525]: Given that this is an internal representation of the mirror config (and that it's shared between infraenv and cluster) I think it should be somewhere internal rather than in the hiveextension API package.

File: internal/controller/controllers/clusterdeployments_controller.go [hunk 9]
```diff
@@ -2325,3 +2343,19 @@ func getClusterDeploymentAdminKubeConfigSecretName(cd *hivev1.ClusterDeployment)
 	}
 	return fmt.Sprintf(adminKubeConfigStringTemplate, cd.Name)
 }
+
+// processMirrorRegistryConfig retrieves the mirror registry configuration from the referenced ConfigMap
+func (r *ClusterDeploymentsReconciler) processMirrorRegistryConfig(ctx context.Context, log logrus.FieldLogger, clusterInstall *hiveext.AgentClusterInstall) (*hiveext.MirrorRegistryConfiguration, error) {
+	mirrorRegistryConfiguration, userTomlConfigMap, err := mirrorregistries.ProcessMirrorRegistryConfig(ctx, log, r.Client, clusterInstall.Spec.MirrorRegistryRef)
```
Comment (carbonin) [inline internal/controller/controllers/clusterdeployments_controller.go:2349]: I'd prefer we read the configmap data in the controllers package here and keep the knowledge of the kube API out of the internal packages.

File: internal/controller/controllers/clusterdeployments_controller_test.go [hunk 10]
```diff
@@ -254,6 +353,293 @@ var _ = Describe("cluster reconcile", func() {
 		mockCtrl.Finish()
 	})
 
+	Context("Mirror Registry", func() {
```
Comment (carbonin) [inline internal/controller/controllers/clusterdeployments_controller_test.go:356]: I'd prefer this be something more like `Describe("processMirrorRegistryConfig", func() {`

That way we can separate the tests that are specifically for this function from the ones that are testing `Reconcile` (which I think should probably be with the other tests).

Generally I don't feel like there's more value in grouping tests by feature rather than the function they're testing.

File: internal/controller/controllers/clusterdeployments_controller_test.go [hunk 11]
```diff
@@ -254,6 +353,293 @@ var _ = Describe("cluster reconcile", func() {
...
+
+			return getTestClusterInstall()
+		}
+
+		createUserMirrorRegistryConfigmap := func(registryToml string, certificate string) {
+			registryCM := newUserProvidedRegistryCM(registryToml, certificate)
+			Expect(c.Create(ctx, registryCM)).ShouldNot(HaveOccurred())
+		}
+
+		When("the user-provided valid image registry contains correct data", func() {
```
Comment (carbonin) [inline internal/controller/controllers/clusterdeployments_controller_test.go:373]: Nit, but single example contexts with no setup code can just inline the context in the test name. I think it will look better in failed test output.

File: internal/controller/controllers/clusterdeployments_controller_test.go [hunk 12]
```diff
@@ -254,6 +353,293 @@ var _ = Describe("cluster reconcile", func() {
...
+			Expect(c.Create(ctx, cluster)).ShouldNot(HaveOccurred())
+			defaultAgentClusterInstallSpec.MirrorRegistryRef = &hiveext.MirrorRegistryConfigMapReference{Name: providedMirrorRegistryCMName, Namespace: testNamespace}
+
+			aci := newAgentClusterInstall(agentClusterInstallName, testNamespace, defaultAgentClusterInstallSpec, cluster)
+			Expect(c.Create(ctx, aci)).ShouldNot(HaveOccurred())
+
+			return cluster, aci
+		}
+
+		getClusterInstallFailure := func(registryToml, certificate string, createMirrorRegistryCM bool) error {
```
Comment (carbonin) [inline internal/controller/controllers/clusterdeployments_controller_test.go:489]: If this is only used in a single context I'd prefer you define it there.

File: internal/controller/controllers/infraenv_controller_test.go [hunk 13]
```diff
@@ -134,6 +135,53 @@ var _ = Describe("infraEnv reconcile", func() {
...
 
+	Context("Mirror registry", func() {
+
+		createUserMirrorRegistryConfigmap := func(registryToml string, certificate string) {
+			registryCM := newUserProvidedRegistryCM(registryToml, certificate)
+			Expect(c.Create(ctx, registryCM)).ShouldNot(HaveOccurred())
+		}
+
+		When("the user-provided valid image registry contains correct data", func() {
+			It("Successfully parse and set the given user configmap", func() {
```
Comment (carbonin) [inline internal/controller/controllers/infraenv_controller_test.go:146]: Since this is all just one big test, it can just be one big test with much less indentation.

File: internal/common/db.go [hunk 14]
```diff
@@ -101,6 +107,27 @@ func (c *Cluster) Payload() any {
...
+func (c *Cluster) GetMirrorRegistryConfiguration() (*v1beta1.MirrorRegistryConfiguration, error) {
+	if c.MirrorRegistryConfiguration == nil {
+		return nil, nil
+	}
+	return convertBytesToMirrorRegistryConfig(c.MirrorRegistryConfiguration)
+}
+
+func (c *Cluster) SetMirrorRegistryConfiguration(configuration *v1beta1.MirrorRegistryConfiguration) error {
+	if configuration == nil {
+		return nil
```
Comment (carbonin) [inline internal/common/db.go:119]: This makes it look like there's no way to explicitly remove the mirror configuration. Do we want to deal with that scenario?

File: internal/ignition/discovery.go [hunk 15]
```diff
@@ -222,6 +223,36 @@ func (ib *ignitionBuilder) shouldAppendOKDFiles(ctx context.Context, infraEnv *c
 	return okdRpmsImage, true
 }
 
+func (ib *ignitionBuilder) setIgnitionMirrorRegistry(ignitionParams *map[string]interface{}, configuration *hiveext.MirrorRegistryConfiguration) error {
+	mirrorRegistriesConfigKey := "MirrorRegistriesConfig"
+	mirrorRegistriesCAConfigKey := "MirrorRegistriesCAConfig"
+
+	if mirrorregistries.IsMirrorConfigurationSet(configuration) {
+		ib.log.Infof("Setting ignition cluster mirror registry to %s", configuration.RegistriesConf)
```
Comment (carbonin) [inline internal/ignition/discovery.go:231]: How large is this string? Should this maybe be a debug message?

File: internal/ignition/discovery.go [hunk 16]
```diff
@@ -325,19 +356,13 @@ func (ib *ignitionBuilder) FormatDiscoveryIgnitionFile(ctx context.Context, infr
...
-			return "", mirrorsErr
-		}
-		ignitionParams["MirrorRegistriesConfig"] = base64.StdEncoding.EncodeToString(registriesContents)
-		ignitionParams["MirrorRegistriesCAConfig"] = base64.StdEncoding.EncodeToString(caContents)
+	mirrorRegistryConfiguration, err := infraEnv.GetMirrorRegistryConfiguration()
+	if err != nil {
+		return "", err
+	}
+
+	if err = ib.setIgnitionMirrorRegistry(&ignitionParams, mirrorRegistryConfiguration); err != nil {
```
Comment (carbonin) [inline internal/ignition/discovery.go:364]: Why pass a reference here? Maps are already references.

https://go.dev/play/p/MOgHmWF1LLp

File: internal/ignition/discovery_test.go [hunk 17]
```diff
@@ -728,6 +730,93 @@ var _ = Describe("IgnitionBuilder", func() {
 		})
 	})
 
+	const (
```
Comment (carbonin) [inline internal/ignition/discovery_test.go:733]: Given that you have a bunch of setup/helpers that are specific to your tests, maybe here is a good place to create a separate context for them.

File: internal/installcfg/builder/builder.go [hunk 18]
```diff
@@ -132,52 +133,93 @@ func (i *installConfigBuilder) getBasicInstallConfig(cluster *common.Cluster) (*
...
-	imageDigestSourceList := make([]installcfg.ImageDigestSource, len(mirrorRegistriesConfigs))
-	for i, mirrorRegistriesConfig := range mirrorRegistriesConfigs {
-		imageDigestSourceList[i] = installcfg.ImageDigestSource{Source: mirrorRegistriesConfig.Location, Mirrors: mirrorRegistriesConfig.Mirror}
+	configuration, err := cluster.GetMirrorRegistryConfiguration()
+	if err != nil {
+		return err
 	}
-	cfg.ImageDigestSources = imageDigestSourceList
+
+	i.log.Infof("found OpenShift version %s, setting mirror registry %+v", cluster.OpenshiftVersion, configuration)
```
Comment (carbonin) [inline internal/installcfg/builder/builder.go:153]: Same here, is this a large string? Maybe debug would be better?

File: internal/installcfg/builder/builder.go [hunk 19]
```diff
@@ -132,52 +133,93 @@ func (i *installConfigBuilder) getBasicInstallConfig(cluster *common.Cluster) (*
...
+
 	return nil
 }
 
-func (i *installConfigBuilder) setImageContentSources(cfg *installcfg.InstallerConfigBaremetal) error {
-	mirrorRegistriesConfigs, err := i.mirrorRegistriesBuilder.ExtractLocationMirrorDataFromRegistries()
-	if err != nil {
-		i.log.WithError(err).Errorf("Failed to get the mirror registries conf need for ImageContentSources")
-		return err
+func (i *installConfigBuilder) setImageDigestMirrorSet(cfg *installcfg.InstallerConfigBaremetal, configuration *v1beta1.MirrorRegistryConfiguration) error {
```
Comment (carbonin) [inline internal/installcfg/builder/builder.go:168]: Is there a way we can deduplicate these functions?

File: pkg/mirrorregistries/mirror_registry.go [hunk 20]
```diff
@@ -0,0 +1,189 @@
+package mirrorregistries
```
Comment (carbonin) [inline pkg/mirrorregistries/mirror_registry.go:1]: Maybe a better file name would be `parser.go`?
Seems a bit weird to just have this be the package name in a different format.

File: internal/bminventory/inventory.go (hunk 2 above)
Comment (eliorerz) [reply internal/bminventory/inventory.go:1675]: Good catch! We also need to set it if it’s nil (in case the mirror configuration was removed)

File: internal/bminventory/inventory_test.go (hunk 3 above)
Comment (eliorerz) [reply internal/bminventory/inventory_test.go:13236]: Sure, added verifications

File: internal/bminventory/inventory_test.go (hunk 4 above)
Comment (eliorerz) [reply internal/bminventory/inventory_test.go:13181]: Added

File: internal/common/db.go (hunk 6 above)
Comment (eliorerz) [reply internal/common/db.go:607]: Yes, it is used in the update cluster

File: internal/common/db.go (hunk 7 above)
Comment (eliorerz) [reply internal/common/db.go:223]: No reason actually, I'll change it to `serialized json` for consistency

File: api/hiveextension/v1beta1/agentclusterinstall_types.go (hunk 8 above)
Comment (eliorerz) [reply api/hiveextension/v1beta1/agentclusterinstall_types.go:525]: Moved to api/common

File: internal/controller/controllers/clusterdeployments_controller.go (hunk 9 above)
Comment (eliorerz) [reply internal/controller/controllers/clusterdeployments_controller.go:2349]: Not sure I understood what you meant here

File: internal/common/db.go (hunk 14 above)
Comment (eliorerz) [reply internal/common/db.go:119]: Fixed that in last commit

File: internal/ignition/discovery_test.go (hunk 17 above)
Comment (eliorerz) [reply internal/ignition/discovery_test.go:733]: It was suupposed to part of `mirror registries config` context. I'll fix taht

File: internal/controller/controllers/clusterdeployments_controller.go (hunk 9 above)
Comment (carbonin) [reply internal/controller/controllers/clusterdeployments_controller.go:2341]: You're sending a configmap reference to the `mirrorregistries` package, but I would expect that package to take the data as input rather than the configmap.

For example `ProcessMirrorRegistryConfig` should also be usable from the REST API functions (if we were ever to implement them)

File: internal/installcfg/builder/builder.go (hunk 19 above)
Comment (eliorerz) [reply internal/installcfg/builder/builder.go]: Not really; both methods, `setImageContentSources` and `setImageDigestMirrorSet` loop over `ImageDigestMirrors` and set different elements within this loop. One sets the `imageContentSourceList`, and the other sets the `imageDigestSourceList`.
If we want to combine them, we would need to loop over this data twice—once for extracting the data and another time for setting it.
In my opinion, this approach is better and more readable than the combined one.

File: api/common/common_types.go [hunk 21]
```diff
@@ -16,3 +18,12 @@ type ValidationsStatus map[string]ValidationResults
 
 // +kubebuilder:object:generate=true
 type ValidationResults []ValidationResult
+
+// MirrorRegistryConfiguration holds the given mirror registry configuration
+type MirrorRegistryConfiguration struct {
```
[the remark this answers is not in the document]
Comment (eliorerz) [reply api/common/common_types.go:23]: Right, when I removed it from aci status it became relevent there

File: internal/controller/controllers/preprovisioningimage_controller.go [hunk 22]
```diff
@@ -548,7 +548,7 @@ func (r *PreprovisioningImageReconciler) AddIronicAgentToInfraEnv(ctx context.Co
 
 	updated := false
 	if string(conf) != infraEnvInternal.InternalIgnitionConfigOverride {
-		_, err = r.Installer.UpdateInfraEnvInternal(ctx, installer.UpdateInfraEnvParams{InfraEnvID: *infraEnvInternal.ID, InfraEnvUpdateParams: &models.InfraEnvUpdateParams{}}, swag.String(string(conf)))
+		_, err = r.Installer.UpdateInfraEnvInternal(ctx, installer.UpdateInfraEnvParams{InfraEnvID: *infraEnvInternal.ID, InfraEnvUpdateParams: &models.InfraEnvUpdateParams{}}, swag.String(string(conf)), nil)
```
[the remark this answers is not in the document]
Comment (eliorerz) [reply internal/controller/controllers/preprovisioningimage_controller.go:551]: In a previous [comment](https://github.com/openshift/assisted-service/pull/6868#discussion_r1819651450), you mentioned that it is safe to remove this logic from here (or did I misunderstand your comment?).

File: internal/ignition/discovery.go [hunk 23]
```diff
@@ -14,9 +14,11 @@ import (
 	"text/template"
 	"time"
 
+	common_api "github.com/openshift/assisted-service/api/common"
 	clusterPkg "github.com/openshift/assisted-service/internal/cluster"
 	"github.com/openshift/assisted-service/internal/common"
 	"github.com/openshift/assisted-service/internal/constants"
+	"github.com/openshift/assisted-service/internal/controller/controllers/mirror_registry"
```
[the remark this answers is not in the document]
Comment (eliorerz) [reply internal/ignition/discovery.go:21]: When I moved `MirrorRegistryConfiguration` to `common/db` I moved this method alongside with it

File: pkg/mirrorregistries/generator.go [hunk 24]
```diff
@@ -78,34 +78,15 @@ func (m *mirrorRegistriesConfigBuilder) ExtractLocationMirrorDataFromRegistries(
...
-		for i := range mirrorTree {
-			currentMirror, ok := mirrorTree[i].Get("location").(string)
-			if !ok {
-				return nil, fmt.Errorf("failed to cast mirror location key to string, registriesConfToml: %s", registriesConfToml)
-			}
-			mirrors = append(mirrors, currentMirror)
-		}
-		registriesConfList[i] = RegistriesConf{Location: location, Mirror: mirrors}
+	for _, i := range idmsMirrors {
+		mirrors := *(*[]string)(unsafe.Pointer(&i.Mirrors))
```
[the remark this answers is not in the document]
Comment (eliorerz) [reply pkg/mirrorregistries/generator.go:88]: As discussed I created a conversion function for that

File: internal/controller/controllers/preprovisioningimage_controller.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (carbonin) [reply internal/controller/controllers/preprovisioningimage_controller.go:551]: That was before we moved the mirror registries to the database.

Ideally it still true, we shouldn't have to care about setting the mirror registries when we're just trying to update the internal override.

I'm not sure how to handle this while also allowing the user to clear these registries. Maybe the internal override and/or the mirror registries should be set with a function separate from this update one?

Alternatively maybe we need some kind of test to make sure `UpdateInfraEnvInternal` doesn't clear the mirror registries when a nil is passed like this if that's the intention (because it sure looks like it will to me).

File: internal/controller/controllers/preprovisioningimage_controller.go [hunk 25]
```diff
@@ -546,9 +547,18 @@ func (r *PreprovisioningImageReconciler) AddIronicAgentToInfraEnv(ctx context.Co
 		return false, err
 	}
 
+	var mirrorRegistryConfiguration *common.MirrorRegistryConfiguration
+	if infraEnvInternal.MirrorRegistryConfiguration != "" {
```
Comment (carbonin) [inline internal/controller/controllers/preprovisioningimage_controller.go:551]: If we're only doing this for the internal ignition override, can we move this into the conditional?

Also generally I'm not thrilled with this approach, but I don't have a better alternative for now without making some significant unrelated changes to how we save the internal ignition so I suppose we'll just leave it for now.

**Comment by eliorerz** [comment]
/override ci/prow/edge-e2e-ai-operator-disconnected-capi 
Due to it's broken
